module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php

```diff
@@ -0,0 +1,819 @@
+     */
+    private function _connectToPatronAPI($patronBarcode, $patronPin = 1234, $forceNewConnection = false)
+    {
+        $patronTokenData = $this->sessionContainer->patronTokenData;
+        $config = $this->getConfig();
+        if ($forceNewConnection || $patronTokenData == null || time() >= $patronTokenData->expirationTime) {
+            $this->debug("connecting to patron API for new token.");
+            $ch = curl_init("https://oauth-patron.overdrive.com/patrontoken");
+            $websiteId = $config->websiteID;
+            $ilsname = $config->ILSname;
+
+            curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 15);
+            curl_setopt($ch, CURLOPT_USERAGENT, "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1)");
+            curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
+            curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
+            curl_setopt($ch, CURLOPT_FOLLOWLOCATION, 1);
+            $encodedAuthValue = base64_encode($config->clientKey . ":" . $config->clientSecret);
+            curl_setopt(
+                $ch,
+                CURLOPT_HTTPHEADER,
+                array(
+                'Content-Type: application/x-www-form-urlencoded;charset=UTF-8',
+                "Authorization: Basic " . $encodedAuthValue,
+                "User-Agent: VuFind-Plus"
+                )
+            );
+
+            curl_setopt($ch, CURLOPT_TIMEOUT, 30);
+            curl_setopt($ch, CURLOPT_POST, 1);
+            
+            //grant_type=password&username=1234567890&password=1234&scope=websiteid:12345 authorizationname:default
+            if ($patronPin == null) {
+                $postFields = "grant_type=password&username={$patronBarcode}&password=ignore&password_required=false&scope=websiteId:{$websiteId}%20authorizationname:{$ilsname}";
+            } else {
+                $postFields = "grant_type=password&username={$patronBarcode}&password={$patronPin}&scope=websiteId:{$websiteId}%20authorizationname:{$ilsname}";
+            }
+            //$this->debug("postFields: $postFields");
+            curl_setopt($ch, CURLOPT_POSTFIELDS, $postFields);
+            curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
+            $return = curl_exec($ch);
+            $curlInfo = curl_getinfo($ch);
+            curl_close($ch);
+            $patronTokenData = json_decode($return);
+            $this->debug("return from OD patron API Call: ". print_r($patronTokenData, true));
+            if (isset($patronTokenData->expires_in)) {
+                $patronTokenData->expirationTime = time() + $patronTokenData->expires_in;
+            } else {
+                $patronTokenData = null;
+            }
+            $this->sessionContainer->patronTokenData = $patronTokenData;
+        }
+        return $patronTokenData;
+    }
+    
+    /**
+     * Call a URL on the API
+     * 
+     * @param string url the url to call
+     *
+     * @return string token for the session
+     */
+    private function _callUrl($url)
```

Comment:
We try to avoid private functions in VuFind, since they limit extensibility. I'd recommend changing this to `protected function callUrl` instead.